### DISCUSSION TITLE
Resolving types using the semantic model

### DISCUSSION THREAD

**Discussion by isolin** [dotnet/roslyn#63094] [Q&A] [answered]
In wider context of my own DSL, I am trying to use Roslyn to resolve types returned by methods or lambda expressions. I am stuck at a few examples I can't solve, also asked a [question on SO](https://stackoverflow.com/questions/73176426/unresolved-return-types-of-lambda-expressions-in-roslyn-semantic-model), but maybe here is a better place to ask.

```cs
public class Wrapper { public Wrapper(int i) { /* ... */} }

var tmp1 = (int y) => Enumerable.Range(0, y).Select(x => new Wrapper(i));
var tmp2 = (int y) => Enumerable.Range(0, y).Select(x => new Wrapper(i)).ToArray();
```

If I compile the code above as `CSharpScript` and get the semantic model of the compilation, it is not able to identify the correct types. A naive approach returns `?` and `TSource[]` respectively:
```cs
//part of a CSharpSyntaxRewriter outlined below
public override SyntaxNode? VisitVariableDeclaration(VariableDeclarationSyntax node)
{
    var info = Semantics.GetSymbolInfo(node.Type);
    var type = info.Symbol;
    Console.WriteLine(type); // SHOULD OUTPUT THE CORRECT TYPE HERE
    return base.VisitVariableDeclaration(node);
}
```

i also tried a more sophisticated approach (or maybe not) tries to get deeper in the syntax tree, but it also fails and by far doesn't cover all cases:
```cs
public class TestRewriter : CSharpSyntaxRewriter
{
    SemanticModel Semantics;
    public TestRewriter(SemanticModel semantics) => Semantics = semantics;

    public override SyntaxNode? VisitParenthesizedLambdaExpression(ParenthesizedLambdaExpressionSyntax node)
    {
        var type = GetMethodReturnType(node);
        Console.WriteLine(type); // SHOULD OUTPUT THE CORRECT TYPE HERE
        return base.VisitParenthesizedLambdaExpression(node);
    }

    private string? GetMethodReturnType(ExpressionSyntax expression)
    {
        if (Semantics.GetSymbolInfo(expression).Symbol is IMethodSymbol methodSymbol)
        {
            var type = methodSymbol.ReturnType.ToString();
            if (( (type?.Contains('?') ?? false)
                  || string.IsNullOrEmpty(methodSymbol.ReturnType.Name))
                && expression is LambdaExpressionSyntax ple)
            {
                if (ple.Body != null)
                {
                    var bodyType = Semantics.GetSymbolInfo(ple.Body);
                    if (bodyType.CandidateReason == CandidateReason.OverloadResolutionFailure
                        && bodyType.CandidateSymbols.Length > 0)
                        && bodyType.CandidateSymbols[0] is IMethodSymbol ms)
                            type = ms.ReturnType.ToString();
                }
                else
                    throw new NotImplementedException(); //should not happen
            }
            return type;
        }
        else if (expression is InvocationExpressionSyntax iexp
                 && iexp.ArgumentList.Arguments.Count > 0)
            return GetMethodReturnType(iexp.ArgumentList.Arguments[0].Expression);
        else
            return null;
    }
}
```

So I am looking for the correct way to resolve the types in the examples correctly, and of course for all other situations where the typing is clear and correct, resolvable at compile time. Since the scripts / programs compile, the type information must be generated somewhere. So I am wondering how to reach it? Maybe compiling a full program instead of a script? Maybe some sort of a recursive search among the symbols? Any help appreciated.

**Comment by CyrusNajmabadi** [comment, marked as answer]
1. always check your COmpilation instance (which you can get from Semantics.Compilation) for diagnostics using .GetDiagnostics() to make everything is ok with your compilation.
2. instead of `var info = Semantics.GetSymbolInfo(node.Type);`, do `var info = Semantics.GetTypeInfo(node.Type);

Let me know if that helps!

**Reply by isolin** [reply]
~~Not at all~~. I checked all that with the debugger before. ~~Diagnostics is empty, all good. The approach in `VisitVariableDeclaration` returns `Func<int, ?>` and `Func<int, TSource[]>` respectively regardless of `GetSymbolInfo` or `GetTypeInfo`.~~

Update: There was a typo, sorry. In `Select` there should be an `x` instead of `i`. Now it works fine.

**Reply by CyrusNajmabadi** [reply]
Did you use .GetDiagnostics()?  I'm not sure what `Diagnostics` is here.

**Comment by isolin** [comment]
Update: My bad! There was a typo. I am wondering why I thought the diagnostics was empty. It works fine now thank you!
